daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala

```diff
@@ -1114,4 +1115,10 @@ trait ElementBase
     case (_, _) => this.subset((floating eq YesNo.No), "Property value floating='yes' is not supported.")
   }
 
+  private lazy val checkInputValueCalcExpression: Boolean = {
+    if (this.inputValueCalcOption.toString.contains("checkConstraints(.)")) {
```

Comment:
So you are depending on the fact that toString of inputValueCalcOption displays the expression. 

The type of inputValueCalcOption is PropertyLookupResult, and that is either a Found or NotFound object. 
(annoying that inputValueCalcOption isn't an Option type, but all the properties work this way.)

I think you should not depend quite so much on the toString of a Found displaying the text of the property. If the property value was say, really big then the toString might elide that with "..." in the textual rep, and that would disable your check. 

Suggest 

```val optIVC = inputValueCalcOption.toOption ```
then use either 
```if (optIVC.isDefined) SDEWhen(optIVC.get.value.contains(...), .... )```
or 
```optIVC.map{ ivc => SDEWhen(ivc.value.contains(...) , ...) }```

Instead of contains, you should compare using a regex match so that "checkConstraints( . )" will still match (has spaces around the ".").  You never know what sort of whitespace convention someone will use for their DFDL schema, particularly for these XPath-like expressions.

Reply:
Ah. I went back and looked at the bug ticket. 

The issue is not about checkConstraints(.), but really any expression involving "." as part of inputValueCalc. The example was calling checkConstraints(.) in an inputValueCalc, but that's just an example of one (rather obscure) use of "." in an inputValueCalc expression. 

To fix this bug really you need to get the compiled expression that is created from the inputValueCalc property, and walk that compiled expression object to see if it has a Self path step reference in it.

Reply:
Of course even that is not entirely correct. A path like "../." which is equivalent to ".." should be allowed. 

Another way to fix this would be to modify the expression compiler with another parameter boolean indicating if the expression is for inputValueCalc. Then for each relative path expresion within the expression, if the relative path is only the Self path, then test the boolean and SDE if true.

Reply:
Does something similar also occur for things other then dfdl:inputValueCalc/checkConstraints? For example:
```xml
<xs:element name="foo" dfdl:length="{ ../foo }"  ... />
```
Are there any DFDL properties where it is valid to access ones self except for checkConstraints? Do we have appropriate errors for those? I guess this issue comes up specificaly with checkConstraints because almost always the argument to checkConstraints?

Reply:
I think the checkConstraints is a red herring. That was just copied out of assert tests into inputValueCalc tests, and so the IVC expression ended up being dfdl:checkConstraints(.). 

The real issue is any use of a "." path in an IVC expression.

Reply:
It feels like maybe sit's more than that? Like an expression can never reference itself except within checkConstraints on a non-inputValueCalc element?

Reply:
To add more complication, is the list of nodes reachable aware of arrays? For example:
```scala
<xs:element name="foo" maxOccurs="unbounded">
  <xs:complexType>
    <xs:element name="bar" type="xs:int" dfdl:length="{ ../../foo[fn:count(../../foo) - 1]/bar }" />
  </xs:complexType>
</xs:element>
```
So ``bar`` can access itself, but only if it is in a previous foo array index. This example feels contrived, but maybe there's an actual use case for something like this?

Reply:
I guess also not preceded nor followed by "," since it could appear as an argument of a multi-arg function call.

Reply:
Feels pretty fragile to me. And is pretty specific to one potential error case. Maybe the current error is actually sufficient? This bug was opened five years ago, maybe our diagnostics surrounding this have improved?

I guess it's better than nothing, but I can imagine building up a bunch of complicated regexes for different conditions that become difficult to maintain and verify correctness.

Reply:
Here's a more realistic example of the mistake we're trying to catch:

```<xs:element name="foo" type="xs:int"
       dfdl:inputValueCalc="{ if (dfdl:checkConstraints(.)) then xs:int(../raw) else -1 }" />
```
The mistake here is they meant to do dfdl:checkConstraints(../raw). The intention here is if the raw value satisfies the facet constraints then use it converting to int. If not substitute -1. 

If we create a test with this mistake in it, and the diagnostic is adequate to figure out what's wrong, then yeah, I'd say this issue is no longer a bug and we don't need an additional compile time check.

Reply:
Is the idea that this inputValueCalc should throw and error? I created a test around it, but no error is thrown.

Reply:
Interesting. Do you have any restrictions on your inputValueCalc element? It's possible that if there are no restrictions checkConstraints doesn't do anything and so it's essentially a no-op that returns true. I'm not sure if that's the expected behavior or not. What happens if you add a restriction?

Reply:
With the above inputValueCalc and restrictions in place the following error is thrown:
Runtime Schema Definition Error: Expression Evaluation Error: Element ex:v does not have a value.

Reply:
I also did a similar test like this:
```xml
<xs:element name="foo" type="xs:int" dfdl:lengthKind="explicit" dfdl:length="{ . }" />
```
And get the same error about ``foo does not have a value``. So this issue of self referential equations produces the same diagnostic. Unfortunately it's a runtime diagnostic, but as discussed above sometimes we can't statically figure it out.

If we do want to improve this diagnostic, perhaps a better route would be to just detect if an expression attempts to access the value of the context from which the expression was run? It's still a runtime check, but it avoids the fragility and specificness of the expression.toString check.

Reply:
In Expression.scala DFDLCheckConstraintsExpr we have the opportunity to examine the args passed to the expression.

In the case of:
dfdl:inputValueCalc="{ if (dfdl:checkConstraints(.)) then xs:int(../ex:cnt) else -1 }" />
The args list is: List(RelativePathExpression(List(Self(None)),false))

In the case of:
dfdl:inputValueCalc="{ if (dfdl:checkConstraints(../.)) then xs:int(../ex:cnt) else -1 }" />
The args list is: List(RelativePathExpression(List(Up(None), Self(None)),false))

Would it be reasonable to detect Self(None) and throw an SDE at compile time?

Reply:
It depends. For example, you could have something like ``dfdl:checkConstraints(./../foo)``, which is weird, but still valid, and has a Self step as the first step. It's possible we collapse steps like this that don't actually do anything, so there might not actually be a Self step in this case. But if there is, then just looking for Self could error on valid expressions.

But it gets even more complicated. Tthe argument to checkConstraints doesn't have to be just a path--it can be an expression to results in a path, so it could be something like this:

```xml
<xs:element name="baz"
  dfdl:inputValueCalc="{ dfdl:checkConstraints(if(../bar eq 1) then ../foo else .) }" />
```

This is invalid, but only if bar is not 1. It would be nice to catch this but this is not trivial to find. You have to realize one of the results of an expression contains self. And as mentioned in other comments, you don't have you use ``.`` to self-refernce. You can step up to parents and then back down to self, and this is difficult to discover.

As I suggested above, I would suggest we just make this a runtime check when trying to access the value of a node. Then we can catch this same self-referencing issue for all expression and not just one tiny little function that's rarely even used. If we're going to do work to fix this issue of a bad diagnostic for self-referencing expressions, let's fix it for everything.

Reply:
In DPath.evaluateExpression: 
<xs:element name="v" type="xs:int" dfdl:lengthKind="explicit" dfdl:length="{.}" />

Is evaluated as:
<CompiledDPath><SelfMove/><IntToLong/></CompiledDPath>

So could be detected in a similar way.

Reply:
Note that this is a very simple example. Most self-referential expressions are going to be more complicated to statically check. And as mentioned before, Self steps are perfectly valid in some cases so doesn't necessarily mean there's a self-reference.

Reply:
Got ya. Thanks for the input.

Reply:
Ok, I have a solution which allows us to detect a null value in the case of a self reference. Does this get us closer?

Reply:
If we can get a diagnostic out that provides reasonable information about what is wrong with the schema, then yes.

Reply:
Wonder if I get a read from you guys on this as a solution before I clean up the tests. This is in UpDownMoves.scala

case object SelfMove extends RecipeOp {
  override def run(dstate: DState): Unit = {
    // do this entirely so it will fail at constant compile time
    // also serves as a sort of assertion check.

    try {
      dstate.currentValue
    } catch {
      case error: Exception =>
        throw new SchemaDefinitionError(None, None, "No value found while evaluating a self referencing expression for element %s.", dstate.currentElement)
    }

    dstate.selfMove()
  }
}

Reply:
So self-move is a no-op if it appears interior to a path, and it is dstate.currentValue if it appears last in a path expression. So, if there is no value for a self move, that has to be "." and that has to be an expression on a simple element. 

Worth a try. You may need to get some context information instead of None, None, so that the user will get the file/line of the expression at least.

Reply:
Thought. A path like ../../foo/bar/. (final dot) isn't a self-reference. So long as that gets converted into just ../../foo/bar (i.e., the self-move operation is removed) then I think we're fine for both parse and unparse. The only place where a self-move would "survive" is in expressions that are just ".", so those are always self-move path expressions.

Reply:
I think a check on ``SelfMove``, while getting closer to a complete solution, still misses a lot of common examples of self referential expressions. For example, it is not uncommon to have an array of integers and maintain a running sum, like this:

```xml
<xs:element name="root">
  <xs:complexType>
    <xs:sequence>
      <xs:element name="array" maxOccurs="unbounded">
        <xs:complexType>
          <xs:sequence>
            <xs:element name="int" type="xs:int" ... />
            <xs:element name="sum" type="xs:int" dfdl:inputValueCalc="{
              if (fn:count(../../array) eq 0)
              then ../int
              else ../int + ../../array[fn:count(../../array) - 1]/sum
            }" />
          </xs:sequence>
        </xs:complexType>
      </xs:element>
    </xs:sequence>
  </xs:complexType>
</xs:element>
```
So each ``sum`` element is an inputValueCalc that adds the current ``int`` element to the previous ``sum`` element.

But if a user accidentally leaves off the `` - 1`` when indexing the array, rather than referencing the previous ``sum`` element the expression would reference the current ``sum`` element, which will be a self referential expression and create the poor diagnostic. So we are doing a self reference without using ``.`` in an expression. I suspect this type of error is very common. (It would probably be worth adding this with the `` - 1`` removed as a test).

----

Thinking more about this, the issue isn't just self-referencing expressions--expressions are actually allowed to self reference it's own **node**. For example, I think it's fine to do something like ``fn:count(.)``, which accesses the node, but not the value (note: I'm not sure if there's a good reason for this type of expression, but I imagine one could come up with reasonble examples). The thing we aren't allowed to self reference is the *value* of a node. Only when you access the value does a self-referencing expression become a problem and need a helpful diagnostic.

And we actually already detect this right now. When a self referencing expression access the value of self, it simply won't have a value yet, and an ``InfosetNoDataException`` is thrown. When this happens when evaluating an expression, the ``handleCompileState`` function in ``DPath.scala`` catches this exception and creates an SDE. This is the diagnostic we ultimately see.

I /believe/ that here is where we need an improved diagnositc. There error we are getting is "element has no data" because this is an ``InfosetNoDataException``, but the underlying reason we have no data here and get an ``InfosetNoDataExcpetion`` in DPath parse must be because it was a self referential expression. I don't think there is any other way this exception can be thrown during parse expression evaluation that is due to something other than a self referencing expression.

Unfortunately, I don't think we can change the message of the ``InfosetNoDataException``, since it is used in other contexts. But I think we can just detect ``InfosetNoDataException`` in the parse part of ``handleCompileState`` and create a new exception with an better diagnostic message (e.g. ``InfosetSelfReferencingException``) and pass this new exception to ``doSDE``.

Note that this does mean we only ever detect this error at runtime, but I with the complexity of expression, detecting it at compile time can be difficult.

Reply:
In the case of <f xsi:nil="true" /> is it valid to throw this error?

Reply:
Does an expression accessing a nilled element also create the ``InfosetNoDataException``? If so, it is definitely incorrect to convert that to a ``SelfReferencingException``. We likely need someone to differentiate the two. Maybe a different exception for trying to access the value of a nilled element is needed, e.g. ``InfosetNilledDataException``. That would probably also improve the dianostics for that error case as well, instead of it being "element has no data" it could be "element is nilled" or something.

Reply:
That said, I'm not sure those limitations aren't just oversights. It may be that these things are needed and will eventually get added. But right now I think there is no notion of nilled and expressions.

Reply:
Looks like the isNilled thing is not a problem.
I pushed the solution Steve indicated for review knowing it will need cleanup is it looks like the way to go.